### PR TITLE
Make input property name changes result in non-incremental task execution

### PR DESCRIPTION
**Pull request by wolfs**
Addresses #4114

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/AbstractNamedFileSnapshotTaskStateChanges.java [hunk 1]
```diff
@@ -58,35 +53,17 @@ protected AbstractNamedFileSnapshotTaskStateChanges(TaskExecution previous, Task
...
         }
         return Iterators.concat(Iterables.transform(getCurrent().entrySet(), new Function<Map.Entry<String, FileCollectionSnapshot>, Iterator<TaskStateChange>>() {
             @Override
             public Iterator<TaskStateChange> apply(Map.Entry<String, FileCollectionSnapshot> entry) {
                 String propertyName = entry.getKey();
                 FileCollectionSnapshot currentSnapshot = entry.getValue();
                 FileCollectionSnapshot previousSnapshot = getPrevious().get(propertyName);
+                if (previousSnapshot == null) {
+                    return Iterators.emptyIterator();
+                }
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/AbstractNamedFileSnapshotTaskStateChanges.java]: Why is this needed now?

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/AbstractPropertyNameTaskStateChanges.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        this.task = task;
+    }
+
+    protected abstract Map<String, ?> getProperties(TaskExecution execution);
+
+    @Override
+    protected void addAllChanges(final List<TaskStateChange> changes) {
+        if (previous == null) {
+            return;
+        }
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/AbstractPropertyNameTaskStateChanges.java]: I wonder why we bother creating any of the `TaskStateChanges` objects if `previous` is `null`...

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/InputFilePropertyNameTaskStateChanges.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+import org.gradle.api.NonNullApi;
+import org.gradle.api.Task;
+import org.gradle.api.internal.changedetection.state.TaskExecution;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+
+@NonNullApi
+public class InputFilePropertyNameTaskStateChanges extends AbstractPropertyNameTaskStateChanges {
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/InputFilePropertyNameTaskStateChanges.java]: It's the _existence,_ not the _name_ of the input file property that changes, no? I'd probably call this `InputFilePropertyTaskStateChanges` instead.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/InputPropertyNameTaskStateChanges.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+import org.gradle.api.NonNullApi;
+import org.gradle.api.Task;
+import org.gradle.api.internal.changedetection.state.TaskExecution;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+
+@NonNullApi
+public class InputPropertyNameTaskStateChanges extends AbstractPropertyNameTaskStateChanges {
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/InputPropertyNameTaskStateChanges.java]: Same here.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/OutputFilePropertyNameTaskChanges.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+import org.gradle.api.NonNullApi;
+import org.gradle.api.Task;
+import org.gradle.api.internal.changedetection.state.TaskExecution;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+
+@NonNullApi
+public class OutputFilePropertyNameTaskChanges extends AbstractPropertyNameTaskStateChanges {
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/OutputFilePropertyNameTaskChanges.java]: And here.

Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/OutputFilePropertyNameTaskChanges.java]: This could even be `OutputPropertyTaskChanges`...

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/InputPropertyNameTaskStateChanges.java (hunk 4 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/InputPropertyNameTaskStateChanges.java]: Can we combine this with `InputFilePropertyTaskStateChanges`?

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/TaskUpToDateState.java [hunk 6]
```diff
@@ -70,7 +75,7 @@ public TaskStateChanges getInputFilesChanges() {
      * Returns if any output files have been changed, added or removed.
      */
     public boolean hasAnyOutputFileChanges() {
-        return outputFileChanges.hasAnyChanges();
+        return noHistoryState.iterator().hasNext() || outputFilePropertyNameChanges.iterator().hasNext() || outputFileChanges.hasAnyChanges();
     }
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/TaskUpToDateState.java]: Why do we need to check for no history here now? If we have to do that, I think we should do it via an explicit method (say, `noHistoryState.hasHistory()` or something), or perhaps just check if `lastExecution == null`.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/AbstractNamedFileSnapshotTaskStateChanges.java (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/AbstractNamedFileSnapshotTaskStateChanges.java]: The check was there before, too, only earlier (before comparing all the properties).

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/AbstractPropertyNameTaskStateChanges.java (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/AbstractPropertyNameTaskStateChanges.java]: I'll look into not creating all those instances when we have no previous execution.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/InputPropertyNameTaskStateChanges.java (hunk 4 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/InputPropertyNameTaskStateChanges.java]: Then we would miss changes like a input file property becoming an input value property, right?

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/TaskUpToDateState.java (hunk 6 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/TaskUpToDateState.java]: Before my change, output file changes was yielding descriptive changes for
- no previous execution
- output file property changes
- changed output files

This is now tracked by different states.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/InputPropertyNameTaskStateChanges.java (hunk 4 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/InputPropertyNameTaskStateChanges.java]: Good point. Damn, this is difficult.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/TaskUpToDateState.java (hunk 6 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/TaskUpToDateState.java]: I see... This might be a little easier to follow I think:

```java
    public boolean hasAnyOutputFileChanges() {
        if (lastExecution == null) {
             return true;
        }
        return outputFilePropertyNameChanges.iterator().hasNext()
            || outputFileChanges.hasAnyChanges();
    }
```

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalTasksIntegrationTest.groovy [hunk 7]
```diff
@@ -354,6 +354,36 @@ ext.added = ['file3.txt', 'file4.txt']
...
+    def "incremental task is informed that all input files are 'out-of-date' when input file property has been added"() {
+        given:
+        file('new-input.txt').text = "new input file"
+        previousExecution()
+
+        when:
+        buildFile << "incremental.inputs.file('new-input.txt')"
+
+        then:
+        executesWithRebuildContext("changed += ['new-input.txt']")
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalTasksIntegrationTest.groovy]: done.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/AbstractPropertyNameTaskStateChanges.java (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/AbstractPropertyNameTaskStateChanges.java]: I made sure that we do not create all those instances when we have no task history. We also need to do much less null checks.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/InputFilePropertyNameTaskStateChanges.java (hunk 3 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/InputFilePropertyNameTaskStateChanges.java]: Renamed.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/OutputFilePropertyNameTaskChanges.java (hunk 5 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/OutputFilePropertyNameTaskChanges.java]: renamed.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/TaskUpToDateState.java (hunk 6 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/TaskUpToDateState.java]: Should be clearer now.

**Comment by wolfs** [comment]
Seems like not evaluating `history.getCurrentExecution()` in `DefaultTaskArtifactStateRepository.getStates()` causes [many test failures](https://builds.gradle.org/viewQueued.html?itemId=10587495) 
https://github.com/gradle/gradle/blob/07859038c5864a9e8476977afb6ac85b01b22311/subprojects/core/src/main/java/org/gradle/api/internal/changedetection/changes/DefaultTaskArtifactStateRepository.java#L201-L213

That was somewhat unexpected for me.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/AbstractNamedFileSnapshotTaskStateChanges.java (hunk 1 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/AbstractNamedFileSnapshotTaskStateChanges.java]: :+1:

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/changes/DefaultTaskArtifactStateRepository.java [hunk 8]
```diff
@@ -201,11 +201,11 @@ private void snapshotAfterOutputsWereGenerated(TaskHistoryRepository.History his
         private TaskUpToDateState getStates() {
             if (states == null) {
                 HistoricalTaskExecution previousExecution = history.getPreviousExecution();
+                CurrentTaskExecution currentExecution = history.getCurrentExecution();
                 if (previousExecution == null) {
                     states = NoHistoryTaskUpToDateState.INSTANCE;
                 } else {
                     // Calculate initial state - note this is potentially expensive
-                    CurrentTaskExecution currentExecution = history.getCurrentExecution();
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/changes/DefaultTaskArtifactStateRepository.java]: Please move the comment, too. We should also mention that resolving current execution is a requirement in a comment.

**Comment by lptr** [comment]
It would be nice to figure out which parts of `getCurrentExecution()` is required, and execute them explicitly instead.

**Comment by wolfs** [comment]
@lptr It seems like we expect every input property to have been evaluated at least once before we start execution. That is why we basically need to call `getCurrentExecution`. Makes sense?

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/changes/DefaultTaskArtifactStateRepository.java (hunk 8 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/changes/DefaultTaskArtifactStateRepository.java]: Moved.

**Comment by wolfs** [comment]
@lptr Please have another look.

**Comment by wolfs** [comment]
@lptr I fixed the performance regression, we are now a little bit [faster than the baseline](https://builds.gradle.org/repository/download/Gradle_Check_PerformanceTestCoordinator/10632336:id/report-performance-performance-tests.zip%21/report/tests/visiting-gzip-tar-trees.html).

Could you have another look?
